Move most of TreeNodeBuilder#build to hash lookup

NickLaMuro opened the pull request:
Purpose
-------
Improve performance building large trees by reducing the cpu overhead required for calls made to each node in the tree.  This should improve performance more noticeably on large trees.


Overview
--------
Large case statements like this require invoking the `===` method, which in some cases can invoke Draper's `===` on presenters, which isn't very cheap since it is coded like this:

```ruby
def ===(other)
  super || (other.respond_to?(:object) && super(other.object))
end
```

The `respond_to?` call is very heavy weight, as it ends up calling `ActiveModel::AttributeMethods#respond_to?` and a bunch deep dives in to the model structure.  Also, the `case` is the equivalent of an `if`/`else`, meaning it does a comparison against the objects it won't even be using.

Since this is just syntactic sugar for saying "is this object is an instance of this class", doing a hash lookup speeds up this process significantly, and there is hardly a change in how the code looks.  The biggest change is in the build method that does the lookup:

```
def build
  instance_exec &BUILD_NODE_HASH[object.class]
```

In which the proc defined in the `BUILD_NODE_HASH` is executed within the context of the current instance.

Decorator classes that might be passed in as `object` in the `#build` method are also handled on class load.  Each key in the `BUILD_NODE_HASH` is looped through to check and see if a corresponding decorator is defined for it.  If it is, the key is added to the hash, and is mapped to the model's node generation proc.


Metrics
-------
On the `/service/explorer/:id` route with about 10k services, this will save about a second (yeah, not amazing, but still)...

**Before**

|      ms | queries | query (ms) |    rows |
|    ---: |    ---: |       ---: |    ---: |
|   19194 |     278 |     1322.2 |   36626 |
|   16158 |     278 |     1280.3 |   36626 |
|   16222 |     278 |     1300.9 |   36626 |
|   15942 |     278 |     1327.6 |   36626 |
|   15904 |     278 |     1291.8 |   36626 |


**After**

|      ms | queries | query (ms) |    rows |
|    ---: |    ---: |       ---: |    ---: |
|   17610 |     278 |     1337.8 |   36626 |
|   15026 |     278 |     1341.9 |   36626 |
|   15344 |     278 |     1292.6 |   36626 |
|   14720 |     278 |     1301.6 |   36626 |
|   14810 |     278 |     1306.2 |   36626 |

These metrics were taken after the changes in https://github.com/ManageIQ/manageiq/pull/11502, but the save from the second even before those changes as well.


Links
----------------
* https://github.com/ManageIQ/manageiq/pull/11502

kbrock commented on app/presenters/tree_node_builder.rb:
```diff
@@ -62,108 +69,126 @@ def build
+  build_node_hash.each do |key|
+    begin
+      new_key = "#{key}Decorator"
+      new_key.constantize # check that it is a valid class
```
Think we want to use `new_key.safe_constantize`
If it is not found, it will return a `nil`.

kbrock replied:
alternatively, we could just remove the "Draper" in `build`.

Either with `sub("Draper","")` or preferably, asking the object if it is a draper object and could we please have the root object.

NickLaMuro replied:
I like both of those ideas better than what I did.  Probably will go with the second since I can just assign the constant in one go then.

kbrock commented on app/presenters/tree_node_builder.rb:
```diff
@@ -62,108 +69,126 @@ def build
+
+  def build
+    # Execute the proc from the BUILD_NODE_HASH in the context of the instance
+    instance_exec(&BUILD_NODE_HASH[object.class.to_s])
```
I'm thinking we should call this with base class.

At first I thought it would lookup class then base class, but now I'm thinking just base class.

```ruby
# possibly
proc = BUILD_NODE_HASH[object.class.to_s] || BUILD_NODE_HASH[object.class.base_class.to_s]
instance_exec(&proc)

# more probably
instance_exec(&BUILD_NODE_HASH[object.class.base_class.to_s])
```

NickLaMuro replied:
Mmmk, sure, I will take a look into that.  Thanks!

kbrock replied:
I'm surprised the `Ems` has was needed for the old code.

For this approach, I think it needs to be base class:
which is a bummer

```ruby
object = object.object if object.kind_of?(Draper::Decorator)
class_name = object.class.base_name.to_s
instance_exec(&BUILD_NODE_HASH[object.class.base_class.to_s])
```

NickLaMuro replied:
Handled this with `base_class` as suggested, but `base_name` wasn't a thing, so I hacked something up that was similar using split.  Seems to not slow things down too much.

kbrock replied:
sorry. I meant `base_class.name`

kbrock commented on app/presenters/tree_node_builder.rb:
```diff
@@ -62,108 +69,126 @@ def build
+    "VmdbIndex"                  => -> { generic_node(object.name, "vmdbindex.png") },
+    "Zone"                       => -> { zone_node },
+    "Hash"                       => -> { hash_node },
+  }
```
think rubocop would have asked for a `.freeze` here

NickLaMuro replied:
Nope, made it `build_node_hash` (local var) here since it was being manipulated in the next `.each`, and assign it as a constnat with `.freeze` it after that.  My way of getting around rubocop, but the `.sub` suggestion should solve that (hopefully), and allow us to assign as a constant from the get go.

kbrock commented on app/presenters/tree_node_builder.rb:
```diff
@@ -62,108 +69,126 @@ def build
 
-    when Vm                   then vm_node(object)
-    when Lan                  then generic_node(object.name,
+    "Vm"                         => -> { vm_node(object) },
```
Yea, I don't think `"Vm"` will get hit.
Maybe do the Ems trick with `VmOrTemplate` and the inner case statement.

NickLaMuro replied:
I guess I don't follow with this one.  Is this something that is failing in the tests (haven't looked at them yet)?  didn't really change anything here and the only thing I could think is a problem is `obeject.class.to_s` never maps to `"Vm"`.  Is that what you meant?

kbrock replied:
```ruby
Vm.first.class.to_s
# => "ManageIQ::Providers::Vmware::InfraManager::Vm"
```

kbrock commented on app/presenters/tree_node_builder.rb:
```diff
@@ -62,108 +69,126 @@ def build
+    "MiqPolicy"                  => -> { miq_policy_node },
+    "MiqPolicySet"               => -> { generic_node(object.description, "policy_profile#{object.active? ? "" : "_inactive"}.png") },
+    "MiqUserRole"                => -> { generic_node(object.name, "miq_user_role.png") },
+    "OrchestrationTemplateCfn"   => -> { generic_node(object.name, "orchestration_template_cfn.png") },
```
possibly make these a single node (for base class) and lookup the png name in a hash / case statement like we do in Ems?

NickLaMuro replied:
Yeah, if we go with the base class, we would probably have to look into changing a lot of what is being done here, which based on the rubocop comment, was not what I was going for here if I could avoid it.

kbrock replied:
nice job avoiding using `base_class` here.

kbrock commented:
just looked at the build failures, does look like it is not finding some of these nodes. think it has to do with the base class / hierarchy lookup.

kbrock commented:
lots of promise. Is a wip for the short time being

NickLaMuro commented:
@kbrock I think I have addressed your concerns with these most recent changes (will add a diff here for the important parts that changed before the rebase):

```diff
diff --git a/app/presenters/tree_node_builder.rb b/app/presenters/tree_node_builder.rb
index e830ff9..c557eed 100644
--- a/app/presenters/tree_node_builder.rb
+++ b/app/presenters/tree_node_builder.rb
@@ -44,9 +44,6 @@ class TreeNodeBuilder
     object.kind_of?(Hash) ? build_hash_id : build_object_id
   end

-  # Long Hash Key is long...
-  MIQ_ORCHESTRATION_TMPL_KEY = "ManageIQ::Providers::Vmware::CloudManager::OrchestrationTemplate".freeze
-
   # FIXME: This is a rubocop disaster... fixed the alignment with the params,
   # but that is about as far as I was willing to go with this one...
   #
@@ -140,7 +137,7 @@ class TreeNodeBuilder
     "OrchestrationTemplateHot"   => -> { generic_node(object.name, "orchestration_template_hot.png") },
     "OrchestrationTemplateAzure" => -> { generic_node(object.name, "orchestration_template_azure.png") },
     "OrchestrationTemplateVnfd"  => -> { generic_node(object.name, "orchestration_template_vnfd.png") },
-    MIQ_ORCHESTRATION_TMPL_KEY   => -> { generic_node(object.name, "orchestration_template_vapp.png") },
+    "OrchestrationTemplate"      => -> { generic_node(object.name, "orchestration_template_vapp.png") },
     "PxeImage"                   => -> { generic_node(object.name, object.default_for_windows ? "win32service.png" : "pxeimage.png") },
     "WindowsImage"               => -> { generic_node(object.name, "os-windows_generic.png") },
     "PxeImageType"               => -> { generic_node(object.name, "pxeimagetype.png") },
@@ -172,22 +169,20 @@ class TreeNodeBuilder
   }
   # rubocop:enable all

-  build_node_hash.each do |key|
-    begin
-      new_key = "#{key}Decorator"
-      new_key.constantize # check that it is a valid class
-      build_node_hash[new_key] = build_node_hash[key]
-
-    # Is there are better way to do this?
-    rescue NameError # rubocop:disable Lint/HandleExceptions
-    end
-  end
-
   BUILD_NODE_HASH = build_node_hash.freeze

   def build
+    # If this is a Decorator, then move grab it's `.object` since we will want
+    # to lookup based on that.
+    obj         =   object.kind_of?(Draper::Decorator) ? object.object : object
+
+    # Find the proc for the class either based on it's non-namespaced class
+    # name, or it's `base_class` (the top level of the inheritence tree)
+    node_lambda =   BUILD_NODE_HASH[obj.class.to_s.split("::").last]
+    node_lambda ||= BUILD_NODE_HASH[obj.class.base_class.to_s]
+
     # Execute the proc from the BUILD_NODE_HASH in the context of the instance
-    instance_exec(&BUILD_NODE_HASH[object.class.to_s])
+    instance_exec(&node_lambda) if node_lambda
     @node
   end
```

NickLaMuro commented on app/presenters/tree_node_builder.rb:
```diff
@@ -44,16 +44,20 @@ def build_id
-    when AssignedServerRole   then assigned_server_role_node(object)
-    when AvailabilityZone     then generic_node(object.name,
+  # FIXME: This is a rubocop disaster... fixed the alignment with the params,
+  # but that is about as far as I was willing to go with this one...
```
Arguably a `WONTFIX` as well, but basically this will require a major re-thinking of how this hash is structured, possibly adding methods to reduce the line length and the nasty procs required here.  Aligning the params was easy, but beyond that the refactoring required to make this pass rubocop normally was out of scope for this PR.  Disabling these cops for the hash instantiation is just the easiest way for now to avoid about 80 cops from blasting this PR, and future PRs that modify this hash.

kbrock commented:
This looks good to me.

1. This is only giving an 8% speed up?
I really expected better results.
Especially going from a long case statement to a hash

Maybe the `instance_eval` is slowing things down? I seem some stuff about it back in the 1.8 days [[ref]](https://nach-vorne.eu/2007/4/18/performance-of-dynamic-code-invokation/) but the VMs are so different now.

Could a proc work? Passing in object or self?

2. Sorry to suggest this...

What kind of performance gains do you get from simply adding the anti draper clause at the top of the old function?

```ruby
object = object.object if object.kind_of?(Draper::Decorator)
```

If it is comparable to this solution, then I'd suggest looking at changing the `instance_eval`.

NickLaMuro commented:
@kbrock In response to your thoughts:

> A. This is only giving an 8% speed up?
I really expected better results.

Valid, though, keep in mind, this is a second regardless of how fast the page is.  So it is 1-2 second savings (8%) when we start from 16 seconds (not great) on the `/service/explorer/:id` page, but also that much when on the `/service/explorer` index page, which is now sitting at about 5ish seconds without this, so like 20%.

I don't think it is the proc, as it should just be evaluating a block within the context of the instance (instead of the class, which it would do if I just did `node_lambda.call`), but I could be wrong.  But this is also related to your next question...

> B. What kind of performance gains do you get from simply adding the anti draper clause at the top of the old function?

This is a good idea, and I will give it a shot.  This might do what we need alone, but the `responds_to?` might still get fired.  Regardless, the minor code change for testing that is too good not to test, so I will get back to you with those results.

NickLaMuro commented:
@kbrock   Here is the follow up metrics for each of the different implementations (this is 10 sequential requests to the `/service/explorer/:id` with a services that has 1849 `.vms`):

### Before

|      ms | queries | query (ms) |    rows |
|    ---: |    ---: |       ---: |    ---: |
|   20456 |     277 |     1361.1 |   36409 |
|   15954 |     277 |     1272.1 |   36409 |
|   16152 |     277 |     1265.7 |   36409 |
|   16038 |     277 |     1284.3 |   36409 |
|   16822 |     277 |     1332.5 |   36409 |
|   16300 |     277 |     1355.8 |   36409 |
|   16639 |     277 |     1370.3 |   36409 |
|   16667 |     277 |     1280.1 |   36409 |
|   15873 |     277 |     1287.2 |   36409 |
|   15846 |     277 |     1264.8 |   36409 |


### Only `object.kind_of?(Draper::Decorator)` patch

Updated the case statement to include the following:


```diff
-  case object
+  obj = object.kind_of?(Draper::Decorator) ? object.object : object
+  case obj
```

|      ms | queries | query (ms) |    rows |
|    ---: |    ---: |       ---: |    ---: |
|   20373 |     277 |     1343.2 |   36409 |
|   16406 |     277 |     1380.1 |   36409 |
|   15915 |     277 |     1271.9 |   36409 |
|   16086 |     277 |     1255.1 |   36409 |
|   15931 |     277 |     1266.2 |   36409 |
|   15627 |     277 |     1305.6 |   36409 |
|   15916 |     277 |     1295.8 |   36409 |
|   15790 |     277 |     1249.9 |   36409 |
|   15479 |     277 |     1263.6 |   36409 |
|   15883 |     277 |     1271.0 |   36409 |


### With PR #11604

|      ms | queries | query (ms) |    rows |
|    ---: |    ---: |       ---: |    ---: |
|   17202 |     277 |     1302.2 |   36409 |
|   14388 |     277 |     1262.4 |   36409 |
|   14815 |     277 |     1274.1 |   36409 |
|   14087 |     277 |     1254.0 |   36409 |
|   14160 |     277 |     1260.1 |   36409 |
|   14293 |     277 |     1262.1 |   36409 |
|   14705 |     277 |     1258.8 |   36409 |
|   14131 |     277 |     1259.7 |   36409 |
|   14407 |     277 |     1268.1 |   36409 |
|   14257 |     277 |     1276.4 |   36409 |




The stackprof summary output after the first patch shows that `ActiveRecord::Core::ClassMethods#===` was still showing up 1.6% of the time, and was 7th on the list for samples.  So having the `case` for this still attributes some loss in performance.

kbrock commented:
ok, I'm still :+1: on this.
@martinpovolny please share any concerns you may have

NickLaMuro commented:
@miq-bot remove_label wip

martinpovolny commented:
@NickLaMuro, @kbrock : euwe?

NickLaMuro commented:
@martinpovolny Had a conversation with @kbrock about this offline, and we are both kind of on the fence about it, so probably leaning to `euwe/no`.  Here is some things to consider:

* If `euwe/no`, any new nodes added to this hash now that are `euwe/yes` will need to do two PRs for it (and could cause some issues):
  1. To work with the hash version in master
  2. To work with the `case` version in `euwe`
* While tests are in place that caught a lot of things that I missed the first pass on this, I am still concerned there are some nodes are a part of this hash that used to work that now don't since we probably don't have a test case for everything there.  I don't think there is anything that was in the case statement that shouldn't **ever** work, but there was a lot of ground to cover so some followup bug fix PRs will probably be necessary.
* The speed updates are nice and should affect all large trees on the backend, but probably at the cost of stability for a bit.

Typing those out probably convinces me that `euwe/no` makes the most sense, but anyone working on the `app/presenters/tree_node_builder.rb` file (especially the UI team, since that is a hot spot for them), and the mergers should be aware of the first point when backporting or considering backporting.

chessbyte commented:
Euwe Backport details:
```sh
$ git log -1
commit 58805a9b65338f6c6c2b632581d06b290f261ad1
Author: Martin Povolny <mpovolny@redhat.com>
Date:   Thu Oct 6 10:37:56 2016 +0200

    Merge pull request #11604 from NickLaMuro/tree_node_builder_instance_build_as_hash
    
    Move most of TreeNodeBuilder#build to hash lookup
    (cherry picked from commit 10f6d9d43cb35b3e6ec7c33f22aae4b24658d318)
```
